### PR TITLE
Feature/open producten integration

### PR DESCRIPTION
**Pull request by Floris272**
Adds Open Producten to Open Inwoner. 

- Added OpenProductenConfig
- Added Category & ProductType Importer classes
- Added task to import data from OP into OI
- uses the existing PDC models to store imported data
  - added `open_producten_uuid` to pdc_models using mixin
  - added Price and Price Option models and minor frontend changes.

### CODE REVIEW

File: src/open_inwoner/openproducten/clients.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    def get_file(self, url) -> File | None:
+        if not url:
+            return None
+
+        file = self._fetch_file(url)
+
+        if not file:
+            return None
+
+        return File.objects.create(original_filename=url.split("/")[-1], file=file)
```
Comment (swrichards) [inline src/open_inwoner/openproducten/clients.py]: Je zou kunnen zeggen dat deze methods hier niet thuis horen. Vanuit de "separation of concerns" gedacht, zou je zeggen dat de client uitsluitend bezig is met dingen uit de API halen. Vervolgens iets _doen_ met die gegevens gebeurt dan in-context. Met andere woorden, deze methodes zijn niet "algemeen" (in de context van de client) maar voor een specifieke context bedoeld (importeren naar  filer). Ik zou deze methodes in de import module plaatsen.

File: src/open_inwoner/openproducten/producttypes_imports.py [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+class ProductTypeImporter(OpenProductenImporter):
+    def __init__(self, client):
+        super().__init__(client)
+        self.product_types = None
+        self.handled_product_types = set()
+
+    @transaction.atomic()
+    def import_producttypes(self):
+        """
+        generate a Product for every ProductType in the Open Producten API
```
Comment (swrichards) [inline src/open_inwoner/openproducten/producttypes_imports.py]: Conventie voor docstrings is volledige zinnen (caps, leestekens):

```suggestion
        Generate a Product for every ProductType in the Open Producten API.
```

File: src/open_inwoner/pdc/models/product.py
Comment (swrichards) [inline src/open_inwoner/pdc/models/product.py:1]: Voor de volledigheid: een belangrijke vraag hier is: willen we ondersteuning behouden voor het scenario waarin je geen OpenProducten gebruikt? (In de praktijk is het antwoord vrijwel zeker ja).

Dat is wat meer werk: dan wil je de mapping waarschijnlijk apart houden (geen OpenProduct mixin, maar aparte models die mappen van de OIP models naar OpenProduct UUIDs). Gezien het hier denk ik vooral om de proof-of-concept gaat, is dit pad voor nu prima, maar iets om ons bewust van te zijn.

File: src/open_inwoner/openproducten/producttypes_imports.py [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+    if commit:
+        model.save()
+
+
+def _get_instance(model, uuid):
+    """Returns an model instance by uuid or None"""
+    return model.objects.filter(open_producten_uuid=uuid).first()
+
+
+class OpenProductenImporter:
```
Comment (swrichards) [inline src/open_inwoner/openproducten/producttypes_imports.py]: Je kan hier [een abstract base class](https://docs.python.org/3.11/library/abc.html) van maken, gezien het niet de bedoeling is dat deze direct geinstantieerd wordt.

File: src/open_inwoner/openproducten/producttypes_imports.py
Comment (swrichards) [inline src/open_inwoner/openproducten/producttypes_imports.py:1]: In deze functies zouden type hints voor de argumenten de leesbaarheid vergroten, om makkelijker te zien waar je Django models verwacht en waar API models.

File: src/open_inwoner/openproducten/producttypes_imports.py [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+                "open_producten_uuid": question.id,
+                "question": question.question,
+                "answer": question.answer,
+                relation_key: relation_object,
+            },
+        )
+        self._add_to_log_list(question_instance, created)
+
+
+class ProductTypeImporter(OpenProductenImporter):
```
Comment (swrichards) [inline src/open_inwoner/openproducten/producttypes_imports.py]: Aanpak is hier op zich prima. Er is sprake van recursiviteit (de product type relations maken zelf weer product types aan) en aardig wat nesting, bij grote datasets zou dat problematisch kunnen worden maar we hebben het hier niet over duizenden rows.

File: src/open_inwoner/openproducten/producttypes_imports.py [hunk 5]
```diff
@@ -0,0 +1,380 @@
...
+                )
+
+                self._handle_product_type(related_product_type)
+
+            related_product_instance = ProductType.objects.get(
+                open_producten_uuid=related_product_type_uuid
+            )
+            product_type_instance.related_products.add(related_product_instance.id)
+
+    def _handle_relations(self, product_type, product_type_instance):
```
Comment (swrichards) [inline src/open_inwoner/openproducten/producttypes_imports.py]: De enige gedachte die me bekruipt bij de recursiviteit en de management van relations is of er gevallen denkbaar zijn waar waar twee product types naar elkaar verwijzen, en is er dan een gevaar voor een inifnite loop c.q. recursion? (Alternatief zou zijn om _eerst_ alle `ProductType` objecten aan te maken met uuid en dummy velden, en dan pas de inhoudelijke veleden te syncen). Maar goed, dat is wellicht een edge case waar we nu nog geen rekening mee hoeven te houden (maar zou een aardige test case zijn).

File: src/open_inwoner/openproducten/producttypes_imports.py [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+        self.product_types = None
+        self.handled_product_types = set()
+
+    @transaction.atomic()
+    def import_producttypes(self):
+        """
+        generate a Product for every ProductType in the Open Producten API
+        """
+
+        self.product_types = self.client.fetch_producttypes_no_cache()
```
Comment (swrichards) [inline src/open_inwoner/openproducten/producttypes_imports.py:114]: Geen pagination?

**Review by swrichards — CHANGES_REQUESTED** [review]
Ziet er goed uit. We kunnen het misschien nog wel even over de aanpak hebben (integratie met bestaande models versus een aparte mapping), maar dat heeft niet blockend te zijn.

File: src/open_inwoner/pdc/models/product.py
Comment (Floris272) [reply src/open_inwoner/pdc/models/product.py:1]: Besproken, mapping zal mogelijk nog worden toegevoegd mocht er tijd over zijn.

File: src/open_inwoner/openproducten/producttypes_imports.py (hunk 5 above)
Comment (Floris272) [reply src/open_inwoner/openproducten/producttypes_imports.py]: Afgehandelde product typen worden bijgehouden en worden niet nog een keer geupdate. Zal een test aan de volgende pr (met alle testen) toevoegen.

File: src/open_inwoner/openproducten/clients.py (hunk 1 above)
Comment (swrichards) [reply src/open_inwoner/openproducten/clients.py]: Ik blijf het niet helemaal correct vinden dat deze functies hier zitten:

1. Er zit weinig client-specifieke functionaliteit in (de URL komt van de parameters, het fetchen van een URL kan ook wherever het nodig is)
2. Er zijn side-effects: er worden models aangemaakt, dat lijkt me niet passen voor een client.

File: src/open_inwoner/openproducten/producttypes_imports.py (hunk 3 above)
Comment (swrichards) [reply src/open_inwoner/openproducten/producttypes_imports.py]: Correctie op mijn eerdere comment: het is niet echt een abc, want de child classes implementeren geen parent methods, maar meer een `Mixin`. Ik denk dat `OpenProductImporterMixin` een betere naam zou zijn, omdat je daarmee aangeeft dat deze class niet bedoeld is om onafhankelijk te instantieren.

**Review by swrichards — CHANGES_REQUESTED** [review]

File: src/open_inwoner/openproducten/clients.py (hunk 1 above)
Comment (Floris272) [reply src/open_inwoner/openproducten/clients.py]: get_image en get_file zijn verplaatst naar de OpenProductenImporter in [567e7e4](https://github.com/Floris272/open-inwoner/pull/2/commits/567e7e4035fdbb3d1c0cff84e65e027619ded420)
